pilot/pkg/proxy/envoy/v2/discovery.go

```diff
@@ -377,48 +374,11 @@ func (s *DiscoveryServer) ConfigUpdate(full bool) {
 // It ensures that at minimum minQuiet time has elapsed since the last event before processing it.
 // It also ensures that at most maxDelay is elapsed between receiving an event and processing it.
 func (s *DiscoveryServer) handleUpdates(stopCh <-chan struct{}) {
-	var timeChan <-chan time.Time
-	var startDebounce time.Time
-	var lastConfigUpdateTime time.Time
-
-	pushCounter := 0
-
-	debouncedEvents := 0
-	fullPush := false
 
 	for {
 		select {
-		case r := <-s.updateChannel:
-			lastConfigUpdateTime = time.Now()
-			if debouncedEvents == 0 {
-				timeChan = time.After(DebounceAfter)
-				startDebounce = lastConfigUpdateTime
-			}
-			debouncedEvents++
-			// fullPush is sticky if any debounced event requires a fullPush
-			if r.full {
-				fullPush = true
-			}
-
-		case now := <-timeChan:
-			timeChan = nil
-
-			eventDelay := now.Sub(startDebounce)
-			quietTime := now.Sub(lastConfigUpdateTime)
-			// it has been too long or quiet enough
-			if eventDelay >= DebounceMax || quietTime >= DebounceAfter {
-				pushCounter++
-				adsLog.Infof("Push debounce stable[%d] %d: %v since last change, %v since last push, full=%v",
-					pushCounter, debouncedEvents,
-					quietTime, eventDelay, fullPush)
-
-				go s.doPush(fullPush)
-				fullPush = false
-				debouncedEvents = 0
-				continue
-			}
-
-			timeChan = time.After(DebounceAfter - quietTime)
+		case f := <-s.updateChannel.C:
```

Comment:
How bout expose a Get() or Dequeue method in  Queue, so no need operate two tricky channels(C and c)

Reply:
We have a public channel and a private channel. The public channel is read-only, but references the same channel instance. We copied this pattern form golang https://golang.org/src/time/tick.go (`Ticker` additionally wraps the private channel inside a `runtimeTimer`, but otherwise it works the same way.)

Is there even way to hide this behind a function call and still be able to use `select`?

Reply:
I would pass the stop channel to the queue, so check stop there.

Reply:
Sorry, I don't understand this. Stopping the queue has nothing to do with processing events stored in a queue.